hs_file_types/utils.py

```diff
@@ -309,22 +324,42 @@ def identify_metadata_files(files):
     return res_files, meta_files, map_files
 
 
-def ingest_logical_file_metadata(metadata_file, resource, map_files):
+def ingest_logical_file_metadata(metadata, resource, map_files):
+    """
+    Ingest logical file metadata into Django models
+    :param metadata: a dict object that contains updated metadata element or an uploaded metadata file object
+    :param resource: the resource object to update metadata for
+    :param map_files: uploaded resource map file object. Default is None for metadata update only
+    :return:
+    """
     resource.refresh_from_db()
-    graph = Graph()
-    graph = graph.parse(data=metadata_file.read())
     agg_type_name = None
+    # if metadata is a dict object with updated metadata, create a rdf graph instance directly from it
+    if isinstance(metadata, dict):
```

Comment:
@aaraney pointed out to me that Python3 has an overload decorator for overloading functions with differently typed parameters.  I'd love to see HydroShare start doing that over checking the types with if/else.  https://docs.python.org/3/library/typing.html#typing.overload

Reply:
Interesting. I'll try the overload decorator in this case.

Reply:
Im not sure that the overload decorator will solve this use case. From my reading, it is an alternative to a `Union` type. I think what will be a better solution in the future will be to use python `3.10`'s [structured pattern matching](https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching). It's basically a switch statement on steroids, very similar to rust's `match` statement. 

@sblack-usu if type checking inputs is so common in HydroShare, we could always write a generic factory library to handle this use case. I'd be happy to chip in to make that happen. Ive written similar code for other projects in the past.

Reply:
I'm not a fan of switch statements, at least for a case like this.  Factory patterns are a bit heavy, we basically want to accept inputs of different types and normalize them to one type before continuing on with the rest of the function.  I'd use it something like this.  Using this isn't the highest priority, it's just the direction I'm looking and there is no effort to standardize something like this across HydroShare currently.
```
@overload
def ingest_logical_file_metadata(metadata: dict, resource, map_files):
    if metadata['type'] in aggregation_type_to_class:
        aggr_class = aggregation_type_to_class[metadata['type']]
        meta_obj = aggr_class.parse_obj(metadata)
        graph = rdf_graph(meta_obj)
        ingest_logical_file_metadata(graph, resource, map_files)
    else:
        raise Exception("{} aggregation type is not supported".format(metadata['type']))


@overload
def ingest_logical_file_metadata(metadata: File, resource, map_files):
    graph = Graph()
    graph = graph.parse(data=metadata.read())
    ingest_logical_file_metadata(graph, resource, map_files)


def ingest_logical_file_metadata(graph: Graph, resource, map_files):
```

Reply:
I understand the hesitance to switch statements in general, im sure we will cross this road together in the future. So what I shared with you @sblack-usu, the `overload` decorator, will not handle type checking and directing input to the appropriate implementation. It is just for type hinting at the moment. I may try to recreate what you mentioned for fun though, I could see it being a little difficult with the reference counter, but it seems feasible.

Reply:
I read the documentation more closely now that you point that out.  That's disappointing.  So to use this `@overload` decorator, you'd then have to use a switch statement in the last function, which is the only function that exists at runtime.

Reply:
It'd basically be the same as the current implementation with an extra bit of type hinting.

Reply:
Below is an incomplete example I threw together @sblack-usu. It gets across the general idea and shows that it is possible. The way python represents function arguments (positional-only, positional-or-keyword, variadic-positional, keyword-only, variadic-keyword) might make writing a robust solution a little hairy. The below solution only works on the straight and narrow, positional-or-keyword that you only treat as positional.

 
```python
import inspect
from typing import Any, Callable, Dict


class overload:
    # {function_name<str>: { function_signature<tuple<type>>: function<function> }}
    funcs: Dict[str, Dict[inspect.Signature, Callable]] = dict()

    def __init__(self, func: Callable) -> None:
        self.func_name = func_name = func.__name__
        func_sig = inspect.signature(func)

        # function signature types
        func_annot = tuple(ann.annotation for ann in func_sig.parameters.values())

        if func_name not in self.funcs:
            self.funcs[func_name] = dict()

        self.funcs[func_name][func_annot] = func

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        print(f"{self.funcs=}")

        # hashable tuple of function args
        arg_types = tuple(type(arg) for arg in args)
        func_matching_types = self.funcs[self.func_name].get(arg_types, None)

        if func_matching_types is None:
            raise TypeError

        return func_matching_types(*args, **kwargs)


@overload
def func(arg: int) -> None:
    return None


@overload
def func(arg: str) -> str:
    return arg


@overload
def func(arg: bool) -> bool:
    return arg


if __name__ == "__main__":
    assert isinstance(func(1), type(None))
    assert isinstance(func("1"), str)
    assert isinstance(func(True), bool)
```

returns
```shell
self.funcs={'func': {(<class 'int'>,): <function func at 0x7fec355d3160>, (<class 'str'>,): <function func at 0x7fec36cd3d30>, (<class 'bool'>,): <function func at 0x7fec36cd3dc0>}}
self.funcs={'func': {(<class 'int'>,): <function func at 0x7fec355d3160>, (<class 'str'>,): <function func at 0x7fec36cd3d30>, (<class 'bool'>,): <function func at 0x7fec36cd3dc0>}}
self.funcs={'func': {(<class 'int'>,): <function func at 0x7fec355d3160>, (<class 'str'>,): <function func at 0x7fec36cd3d30>, (<class 'bool'>,): <function func at 0x7fec36cd3dc0>}}
```